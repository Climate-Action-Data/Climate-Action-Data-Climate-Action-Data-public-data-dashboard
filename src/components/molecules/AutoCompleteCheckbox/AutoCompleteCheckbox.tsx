import { FC, useState } from 'react'
import { Box, Button, Checkbox, Flex, Input, Popover, PopoverBody, PopoverContent, PopoverHeader, PopoverTrigger, Spacer, Stack, StackDivider, VStack } from '@chakra-ui/react'
import { SearchIcon } from '@/components/atoms/SearchIcon/SearchIcon'
import { DropDownIcon } from '@/components/atoms/DropDownIcon/DropDownIcon'
import { FilterCountIndicator } from '@/components/atoms/FilterCountIndicator/FilterCountIndicator'

interface AutoCompleteCheckboxProps {
  label: string
  noOfSelectedFilters: number
  options: string[]
  selectedFilters: string[]
  applyFilters: (values: string[]) => void
}

const AutoCompleteCheckbox: FC<AutoCompleteCheckboxProps> = (props) => {
  const { label, options, applyFilters, noOfSelectedFilters, selectedFilters } = props
  const [selectedValues, setSelectedValues] = useState<string[]>([])
  const [searchInput, setSearchInput] = useState<string>(``)

  const allChecked = options.length != 0 && selectedValues.length === options.length
  const isIndeterminate = selectedValues.length !== 0 && selectedValues.length !== options.length

  return (
    <Popover
      gutter={0}
      isLazy
      placement="bottom-start"
      matchWidth
      onClose={() => {
        setSearchInput(``)
        setSelectedValues([...selectedFilters])
      }}
    >
      <PopoverTrigger>
        <Button variant={noOfSelectedFilters !== 0 ? `dropdownSelected` : `dropdownUnselected`}>
          <Flex fontFamily={`aeonik`} fontWeight={`normal`} fontSize={`16px`} alignItems={`center`} grow={1}>
            {label}
            <Box width={`8px`} />
            {noOfSelectedFilters !== 0 && <FilterCountIndicator count={noOfSelectedFilters} />}
            <Spacer />
            <DropDownIcon />
          </Flex>
        </Button>
      </PopoverTrigger>
      <PopoverContent border={0} borderRadius={`4px`} width={[`100%`, null, null, `298px`]} backgroundColor={`white`} boxShadow={`xl`}>
        <PopoverHeader border={0} boxShadow="md" padding={0}>
          <VStack divider={<StackDivider height={`1px`} borderColor={`#B8BEC0`} />} spacing={0}>
            <Flex width={`100%`} alignItems={`center`} padding={`8px`}>
              <SearchIcon color={`lightGray.700`} paddingY={`auto`} marginRight={`8px`} />
              <Input
                placeholder={`Search`}
                variant={`autoCompleteCheckboxTextInput`}
                value={searchInput}
                onChange={(event) => {
                  setSearchInput(event.target.value)
                }}
              />
            </Flex>
            <Flex width={`100%`} marginTop={`4px`} alignItems={`center`} padding={`8px 0.75rem`}>
              <Checkbox
                isIndeterminate={isIndeterminate}
                isChecked={allChecked}
                onChange={() => {
                  if (isIndeterminate || selectedValues.length === 0) {
                    setSelectedValues([...options])
                  } else {
                    setSelectedValues([])
                  }
                }}
              />
              <Spacer />
              <Button onClick={() => applyFilters(selectedValues)} variant={`textLink`}>
                Apply
              </Button>
            </Flex>
          </VStack>
        </PopoverHeader>
        <PopoverBody maxHeight={`200px`} width={`100%`} overflowY={`scroll`}>
          <Stack spacing={5} direction="column">
            {options &&
              options.map(
                (value, index) =>
                  value.toUpperCase().includes(searchInput.toUpperCase()) && (
                    <Checkbox
                      key={`${label}-${index}`}
                      isChecked={selectedValues.includes(value)}
                      onChange={(event) => {
                        if (event.target.checked) {
                          setSelectedValues([...selectedValues, value])
                        } else {
                          setSelectedValues((prevState) => prevState.filter((selectedValue) => selectedValue !== value))
                        }
                      }}
                    >
                      {value}
                    </Checkbox>
                  ),
              )}
          </Stack>
        </PopoverBody>
      </PopoverContent>
    </Popover>
  )
}
export default AutoCompleteCheckbox
