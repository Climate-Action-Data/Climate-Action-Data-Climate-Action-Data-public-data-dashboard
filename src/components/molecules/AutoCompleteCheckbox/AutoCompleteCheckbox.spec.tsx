import '@testing-library/jest-dom/extend-expect'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import AutoCompleteCheckbox from '@/components/molecules/AutoCompleteCheckbox/AutoCompleteCheckbox'

describe(`AutoCompleteCheckbox`, () => {
  const tLabel = `search`
  const tOptions = [`option1`, `option2`, `option3`]

  test(`should render AutoCompleteCheckbox with no data input`, () => {
    const { container } = render(<AutoCompleteCheckbox label={tLabel} noOfSelectedFilters={0} options={[]} selectedFilters={[]} applyFilters={() => null} />)
    expect(container).toMatchSnapshot()
  })

  test(`should render AutoCompleteCheckbox with options`, () => {
    const { container } = render(<AutoCompleteCheckbox label={tLabel} noOfSelectedFilters={0} options={tOptions} selectedFilters={[]} applyFilters={() => null} />)
    expect(container).toMatchSnapshot()
  })

  test(`should render AutoCompleteCheckbox with options 2`, async () => {
    const { findByText } = render(<AutoCompleteCheckbox label={tLabel} noOfSelectedFilters={0} options={tOptions} selectedFilters={[]} applyFilters={() => null} />)
    fireEvent.click(await findByText(tLabel))
    expect(findByText(tOptions[0])).toBeDefined()
    expect(findByText(tOptions[1])).toBeDefined()
  })

  test(`should render AutoCompleteCheckbox with options 3`, async () => {
    const { findByText } = render(<AutoCompleteCheckbox label={tLabel} noOfSelectedFilters={1} options={[`option1`]} selectedFilters={[`option1`]} applyFilters={() => null} />)

    expect(findByText(1)).toBeDefined()
  })

  test(`should be able to apply filter`, async () => {
    const tApplyFilters = jest.fn()
    const { findByText, findByRole } = render(<AutoCompleteCheckbox label={tLabel} noOfSelectedFilters={0} options={tOptions} selectedFilters={[]} applyFilters={tApplyFilters} />)
    fireEvent.click(await findByText(tLabel))
    const tOption = await findByRole(`checkbox`, { name: `option1` })
    const tApply = await findByText(`Apply`)
    fireEvent.click(tOption)
    expect(tOption).toBeChecked()
    fireEvent.click(tOption)
    expect(tOption).not.toBeChecked()
    fireEvent.click(tOption)
    fireEvent.click(tApply)
    expect(tApplyFilters).toHaveBeenCalledWith([`option1`])
  })

  test(`should filter based on input`, async () => {
    const { findByPlaceholderText, findByText, queryByText } = render(
      <AutoCompleteCheckbox label={tLabel} noOfSelectedFilters={0} options={tOptions} selectedFilters={[]} applyFilters={() => null} />,
    )
    fireEvent.click(await findByText(tLabel))
    const tSearchField = await findByPlaceholderText(`Search`)
    fireEvent.change(tSearchField, { target: { value: `option2` } })
    const tOption_1 = queryByText(`option1`)
    expect(tOption_1).toBeNull()
  })

  test(`should trigger AutoCompleteCheckbox onClose`, async () => {
    render(<AutoCompleteCheckbox label={tLabel} noOfSelectedFilters={0} options={tOptions} selectedFilters={[]} applyFilters={() => null} />)
    await userEvent.click(await screen.findByText(tLabel))
    await userEvent.click(await screen.findByText(tLabel))
    await waitFor(() => expect(screen.queryByText(tOptions[0])).toBeDefined(), { timeout: 2 })
  })
})
