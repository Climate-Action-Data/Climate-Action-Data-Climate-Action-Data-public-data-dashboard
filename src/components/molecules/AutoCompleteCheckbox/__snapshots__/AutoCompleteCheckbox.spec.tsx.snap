// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoCompleteCheckbox should render AutoCompleteCheckbox with no data input 1`] = `
<div>
  <button
    aria-controls="popover-content-:r1:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="chakra-button css-4xx2wk"
    id="popover-trigger-:r1:"
    type="button"
  >
    <div
      class="css-6t88ve"
    >
      search
      <div
        class="css-1gq8co7"
      />
      <div
        class="css-17xejub"
      />
      <svg
        class="chakra-icon css-c0air1"
        focusable="false"
        viewBox="0 0 8 4"
      >
        <path
          d="M3.99911 3.9034L0.191406 0.0957031H7.80681L3.99911 3.9034Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </button>
  <div
    class="chakra-popover__popper css-0"
    style="visibility: hidden; position: absolute; inset: 0 auto auto 0;"
  >
    <section
      class="chakra-popover__content css-lra071"
      id="popover-content-:r1:"
      role="dialog"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
      tabindex="-1"
    />
  </div>
</div>
`;

exports[`AutoCompleteCheckbox should render AutoCompleteCheckbox with options 1`] = `
<div>
  <button
    aria-controls="popover-content-:r3:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="chakra-button css-4xx2wk"
    id="popover-trigger-:r3:"
    type="button"
  >
    <div
      class="css-6t88ve"
    >
      search
      <div
        class="css-1gq8co7"
      />
      <div
        class="css-17xejub"
      />
      <svg
        class="chakra-icon css-c0air1"
        focusable="false"
        viewBox="0 0 8 4"
      >
        <path
          d="M3.99911 3.9034L0.191406 0.0957031H7.80681L3.99911 3.9034Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </button>
  <div
    class="chakra-popover__popper css-0"
    style="visibility: hidden; position: absolute; inset: 0 auto auto 0;"
  >
    <section
      class="chakra-popover__content css-lra071"
      id="popover-content-:r3:"
      role="dialog"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
      tabindex="-1"
    />
  </div>
</div>
`;
